test/http/http-client.test.js

```diff
@@ -0,0 +1,209 @@
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+const should = require('should');
+
+const { importFile, getTestOptions } = require('../testutils');
+const MockRequest = require('../mock-request');
+
+const HttpClient = importFile('http/http-client');
+const HttpRequest = importFile('http/http-request');
+const DirectBinaryUploadOptions = importFile('direct-binary-upload-options');
+const HttpResult = importFile('http-result');
+
+describe('HTTP Client Tests', function() {
+    let httpResult;
+    let options;
+    let uploadOptions;
+    let httpClient;
+
+    beforeEach(function() {
+        MockRequest.reset();
+        options = getTestOptions();
+        uploadOptions = new DirectBinaryUploadOptions()
+            .withHttpRetryDelay(10);
+        httpClient = new HttpClient(options, uploadOptions);
+        httpResult = new HttpResult(options, uploadOptions);
+    });
+
+    function createHttpRequest(url) {
+        return new HttpRequest(options, url);
+    }
+
+    it('http request submit smoke test', async function() {
+        MockRequest.onGet('http://myunittest.com').reply(200, 'hello world!');
+        const response = await httpClient.submit(createHttpRequest('http://myunittest.com'));
+
+        should(response).be.ok();
+        should(response.getStatusCode()).be.exactly(200);
+        should(response.getData()).be.exactly('hello world!');
+        should(response.getElapsedTime() !== undefined).be.ok();
+    });
+
+    it('test submit retry', async function() {
+        const url = 'http://myretryunittest';
+        MockRequest.onGet(url).replyOnce(500);
+        MockRequest.onGet(url).reply(200, 'retried!');
+
+        const response = await httpClient.submit(createHttpRequest(url), httpResult);
+        should(response).be.ok();
+
+        should(MockRequest.history.get.length).be.exactly(2);
+        should(httpResult.getRetryErrors().length).be.exactly(1);
+    });
+
+    it('test submit retry all fail', async function() {
+        const url = 'http://myfailingretryunittest';
+        MockRequest.onGet(url).reply(500);
+
+        let threw = false;
+        try {
+            await httpClient.submit(createHttpRequest(url));
+        } catch (e) {
+            threw = true;
+        }
+        should(MockRequest.history.get.length).be.exactly(3);
+        should(threw).be.ok();
+    });
+
+    it('test submit retry not retryable error', async function() {
+        const url = 'http://mynotfoundunittest';
```

Comment:
Should HTTPS be supported and tested too?

Reply:
From a unit test perspective, I don't know that this will make a difference. I think this case would be a good candidate for an integration test.

Reply:
In Nui we support only HTTPS and explicitly refuse HTTP. When I look at this test, it seems aem-upload accepts unencrypted HTTP urls?